pkg/provenance/provenance.go

```diff
@@ -147,122 +148,436 @@ func (o ObjectLineage) GetVersions() string {
+			jointQueryResults := make([]bool, len(attributeCombinedQuery))
+			//jointQueryResults is a boolean slice that represents the satisfiability
+			//of the joint query. (all need to be true for it to have found qkey to be true)
+			i := 0
+			for _, pair := range attributeCombinedQuery {
+				qckey := pair[0] //for each field/value pair, must find each qckey in
+												 //the map object mymap
+				qcval := pair[1]
+				for okey, ovalue := range mymap {
+					if qckey == okey && qcval == ovalue {
+						jointQueryResults[i] = true
+					}
+				}
+				i += 1
+			}
+			allTrue := true
+			for _, b := range jointQueryResults {
+				if !b {
+					allTrue = false
+				}
+			}
+			if allTrue{
+				return true //satisfied the joint query
+			}
+		} else {
+			//if there is no attribute relationship, but the mapslice type assert was fine,
+			//only need to find an okey in one of the maps, where that query field/value (qkey,qvalue)
+			//is satisfied.
+			for okey, ovalue := range mymap {
+				if qkey == okey && qval == ovalue {
+					return true
+				}
+			}
+		}
+	}
+	return false
 }
-
 func (o ObjectLineage) FullDiff(vNumStart, vNumEnd int) string {
 	var b strings.Builder
 	sp1 := o[vNumStart]
 	sp2 := o[vNumEnd]
 	for attribute, data1 := range sp1.AttributeToData {
-		if data2, ok := sp2.AttributeToData[attribute]; ok {
-			if data1 != data2 {
-				fmt.Fprintf(&b, "FOUND DIFF")
-				fmt.Fprintf(&b, "Spec version %d:\n %s\n", vNumStart, data1)
-				fmt.Fprintf(&b, "Spec version %d:\n %s\n", vNumEnd, data2)
+		data2, ok := sp2.AttributeToData[attribute] //check if the attribute even exists
+		if ok {
+			str1, ok1 := data1.(string)
+			str2, ok2 := data2.(string)
+			if ok1 && ok2 && str1 != str2 {
+				fmt.Fprintf(&b, "Found diff on attribute %s:\n", attribute)
+				fmt.Fprintf(&b, "\tVersion %d: %s\n", vNumStart, data1)
+				fmt.Fprintf(&b, "\tVersion %d: %s\n", vNumEnd, data2)
 			} else {
-				fmt.Fprintf(&b, "No difference for attribute %s \n", attribute)
+				// fmt.Fprintf(&b, "No difference for attribute %s \n", attribute)
 			}
+			int1, ok1 := data1.(int)
+			int2, ok2 := data2.(int)
+			if ok1 && ok2 && int1 != int2 {
+				fmt.Fprintf(&b, "Found diff on attribute %s:\n", attribute)
+				fmt.Fprintf(&b, "\tVersion %d: %d\n", vNumStart, int1)
+				fmt.Fprintf(&b, "\tVersion %d: %d\n", vNumEnd, int2)
+			} else {
+				// fmt.Fprintf(&b, "No difference for attribute %s \n", attribute)
+			}
+			strArray1, ok1 := data1.([]string)
+			strArray2, ok2 := data2.([]string)
+			if ok1 && ok2 {
+				for _, str := range strArray1 {
+					found := false
+					for _, val := range strArray2 {
+						if str == val {
+							found = true
+						}
+					}
+					if !found { // if an element does not have a match in the next version
+						fmt.Fprintf(&b, "Found diff on attribute %s:\n", attribute)
+						fmt.Fprintf(&b, "\tVersion %d: %s\n", vNumStart, strArray1)
+						fmt.Fprintf(&b, "\tVersion %d: %s\n", vNumEnd, strArray2)
+					}
+				}
```

Comment:
Here there is no else clause -- that makes sense.
We should do same for string and int checks.